File: src/main/scala/li/cil/oc/OpenComputers.scala
package li.cil.oc

import li.cil.oc.common.{IMC, Proxy}
import li.cil.oc.common.block.ChameliumBlock
import li.cil.oc.common.blockentity.BlockEntityTypes
import li.cil.oc.common.datacomponents.OCComponents
import li.cil.oc.common.entity.EntityTypes
import li.cil.oc.common.init.{Blocks, Items}
import li.cil.oc.common.menu.MenuTypes
import li.cil.oc.common.recipe.Recipes
import li.cil.oc.integration.Mods
import li.cil.oc.server.loot.LootFunctions
import li.cil.oc.util.ThreadPoolFactory
import net.neoforged.api.distmarker.Dist
import net.neoforged.bus.api.{IEventBus, SubscribeEvent}
import net.neoforged.fml.{InterModComms, ModContainer}
import net.neoforged.fml.common.Mod
import net.neoforged.fml.event.lifecycle.{FMLCommonSetupEvent, InterModProcessEvent}
import net.neoforged.fml.loading.FMLPaths
import net.neoforged.neoforge.common.NeoForge
import net.neoforged.neoforgespi.Environment
import org.apache.logging.log4j.{LogManager, Logger}

import java.nio.file.Paths
import scala.collection.convert.ImplicitConversionsToScala._

object OpenComputers {
  final val ID = "opencomputers"

  final val Name = "OpenComputers"

  final val McVersion = "1.20.1-forge"

  final val Version = "1.9.0-beta"

  final val log: Logger = LogManager.getLogger(Name)

  var proxy: Proxy = _

  private var instance: Option[OpenComputers] = None

  def get = instance match {
    case Some(oc) => oc
    case _ => throw new IllegalStateException("not initialized")
  }
}

@Mod(OpenComputers.ID)
class OpenComputers(modBus: IEventBus, modContainer: ModContainer) {
  OpenComputers.proxy = {
    val cls = Environment.get.getDist match {
      case Dist.CLIENT => Class.forName("li.cil.oc.client.Proxy")
      case _ => Class.forName("li.cil.oc.common.ServerProxy")
    }
    cls.getConstructor(classOf[IEventBus]).newInstance(modBus).asInstanceOf[Proxy]
  }

  Settings.load(FMLPaths.CONFIGDIR.get().resolve(Paths.get("opencomputers", "settings.conf")).toFile())

  modBus.register(this)
  OCComponents.REGISTRAR.register(modBus)
  Items.init(modBus)
  Blocks.init(modBus)
  CreativeTab.CREATIVE_TABS.register(modBus)
  BlockEntityTypes.init(modBus)
  Recipes.init(modBus)
  LootFunctions.init(modBus)
  EntityTypes.ENTITY_TYPES.register(modBus)
  MenuTypes.MENU.register(modBus)
  //modBus.register(classOf[Capabilities])
  modBus.register(li.cil.oc.data.DataGenerators)
  modBus.register(CreativeTab)
  OpenComputers.instance = Some(this)
  //NeoForge.EVENT_BUS.register(OpenComputers.proxy)
  modBus.register(OpenComputers.proxy)
  OpenComputers.proxy.preInit()
  NeoForge.EVENT_BUS.register(ThreadPoolFactory)

  // these used to use @EventBusSubscriber but Scala makes this impossible on NeoForge
  modBus.register(ChameliumBlock)

  Mods.preInit() // Must happen after loading Settings but before registry events are fired.

  @SubscribeEvent
  def imc(e: InterModProcessEvent): Unit = {
    // Technically requires synchronization because IMC.sendTo doesn't check the loading stage.
    e.enqueueWork((() => {
      InterModComms.getMessages(OpenComputers.ID).sequential.iterator.foreach(IMC.handleMessage)
    }): Runnable)
  }

  @SubscribeEvent
  def onCommonSetup(e: FMLCommonSetupEvent): Unit = {
    OpenComputers.proxy.init(e)
  }
}

File: src/main/kotlin/li/cil/oc/OpenComputers.kt
package li.cil.oc

import net.minecraftforge.eventbus.api.IEventBus
import net.minecraftforge.fml.ModContainer
import net.minecraftforge.fml.ModLoadingContext
import net.minecraftforge.fml.common.Mod
import net.minecraftforge.fml.javafmlmod.FMLJavaModLoadingContext
import net.minecraftforge.network.simple.SimpleChannel
import org.apache.logging.log4j.LogManager
import org.apache.logging.log4j.Logger

@Mod(OpenComputers.ID)
class OpenComputers {
    companion object {
        const val ID = "opencomputers"
        const val Name = "OpenComputers"
        const val McVersion = "1.20.1-forge"
        const val Version = "1.9.0-beta"
        @JvmField val log: Logger = LogManager.getLogger(Name)
        //@JvmField val proxy: Proxy by lazy {
        //    val dist = Environment.get().getDist()
        //    val cls = when (dist) {
        //        Dist.CLIENT -> Class.forName("li.cil.oc.client.Proxy")
        //        else -> Class.forName("li.cil.oc.common.Proxy")
        //    }
        //    cls.getConstructor().newInstance() as Proxy
        //}
    }

    val modContainer: ModContainer = ModLoadingContext.get().getActiveContainer()
    val modBus: IEventBus = FMLJavaModLoadingContext.get().modEventBus
    
    val channel: SimpleChannel? = null
    
    init {
        modBus.register(this)
        /*
        Items.init(modBus)
  Blocks.init(modBus)
  CreativeTab.CREATIVE_TABS.register(modBus)
  TileEntityTypes.init(modBus)
  Recipes.init(modBus)
  LootFunctions.init(modBus)
  EntityTypes.ENTITY_TYPES.register(modBus)
  MenuTypes.MENU_TYPES.register(modBus)
  modBus.register(classOf[Capabilities])
  modBus.register(li.cil.oc.data.DataGenerators)
  modBus.register(CreativeTab)
  OpenComputers.instance = Some(this)
  MinecraftForge.EVENT_BUS.register(OpenComputers.proxy)
  modBus.register(OpenComputers.proxy)
  Settings.load(FMLPaths.CONFIGDIR.get().resolve(Paths.get("opencomputers", "settings.conf")).toFile())
  OpenComputers.proxy.preInit()
  MinecraftForge.EVENT_BUS.register(ThreadPoolFactory)
  Mods.preInit()
         */
    }
}